Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/FlatPlatePaalmanPingsCorrection.py

```diff
@@ -504,83 +552,142 @@ def _flat_abs(self, angle):
-                # transmission case
-                acc2 = sst(ki_c2, kf_c2)
-            else:
-                # reflection case
-                acc2 = ssr(ki_c2, kf_c2)
+    # ------------------------------------------------------------------------------
+
+    def _can_thickness_calc(self, can_x_section, can_x_section_efixed, can_thickness, wavelengths, theta, alpha, stha, salpha, ssr, sst):
+        if self._emode == 'Efixed':
+            ki = can_x_section_efixed * can_thickness / salpha
+            kf = can_x_section_efixed * can_thickness / stha
+        else:
+            ki, kf = self._calc_ki_kf(wavelengths, can_thickness, salpha, stha, can_x_section, can_x_section_efixed)
+
+        if theta < alpha or theta > (alpha + np.pi):
+            # transmission case
+            acc = sst(ki, kf)
+        else:
+            # reflection case
+            acc = ssr(ki, kf)
+
+        return ki, kf, acc
 
-            # Attenuation due to passage by other layers (sample or container)
-            if theta < alpha or theta > (alpha + np.pi):                  # transmission case
-                acc = self._can_front_thickness * acc1 * np.exp(-kf_c2)
-                acc += self._can_back_thickness * acc2 * np.exp(-ki_c1)
-                acc /= (self._can_front_thickness + self._can_back_thickness)
+    # ------------------------------------------------------------------------------
 
-                acsc = self._can_front_thickness * acc1 * np.exp(-kf_s-kf_c2)
-                acsc += self._can_back_thickness * acc2 * np.exp(-ki_c1-ki_s)
+    def _container_transmission_calc(self, acc, acc1, acc2, ki_s, kf_s, ki_c1, kf_c2, ass):
+        if self._has_can_front_in and self._has_can_back_in:
+            acc = self._can_front_thickness * acc1 * np.exp(-kf_c2)
+            acc += self._can_back_thickness * acc2 * np.exp(-ki_c1)
+            acc /= (self._can_front_thickness + self._can_back_thickness)
+        elif self._has_can_front_in:
+            acc = acc1
+        elif self._has_can_back_in:
+            acc = acc2
+
+        if self._has_sample_in:
+            if self._has_can_front_in and self._has_can_back_in:
+                acsc = self._can_front_thickness * acc1 * np.exp(-kf_s - kf_c2)
+                acsc += self._can_back_thickness * acc2 * np.exp(-ki_c1 - ki_s)
                 acsc /= (self._can_front_thickness + self._can_back_thickness)
+            elif self._has_can_front_in:
+                acsc = acc1 * np.exp(-kf_s)
+            elif self._has_can_back_in:
+                acsc = acc2 * np.exp(-ki_s)
+            else:
+                acsc = 0.5 * np.exp(-kf_s) + 0.5 * np.exp(-ki_s)
+        else:
+            acsc = acc
+
+        if self._has_can_front_in and self._has_can_back_in:
+            assc = ass * np.exp(-ki_c1 - kf_c2)
+        elif self._has_can_front_in:
+            assc = ass * np.exp(-ki_c1)
+        elif self._has_can_back_in:
+            assc = ass * np.exp(-kf_c2)
```

Comment:
np.exp(-ki_c1) and np.exp(-kf_c2) are being calculated for the second time in this function, are these factors likely to be small in size?

Reply:
This should not actually get called twice, if `self._has_can_front_in` and  `self._has_can_back_in` then `np.exp(-kf_c2)` is called on line 622, if only `self._has_can_back_in` then it is only called on line 649. I agree though it is a bit hard to follow this, so I re-arranged these methods. I hope they read a bit more clearly now.